ext4: start searching for the right extent from the goal group.

From: "Aneesh Kumar K.V" <aneesh.kumar@linux.vnet.ibm.com>

With mballoc we search for the best extent using different
criteria. We should always use the goal group when we are
starting with a new criteria.

Signed-off-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
Signed-off-by: Mingming Cao <cmm@us.ibm.com>
Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
---
 fs/ext4/mballoc.c |   10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

Index: linux-2.6.26-rc5/fs/ext4/mballoc.c
===================================================================
--- linux-2.6.26-rc5.orig/fs/ext4/mballoc.c	2008-06-06 17:03:40.000000000 -0700
+++ linux-2.6.26-rc5/fs/ext4/mballoc.c	2008-06-06 17:03:43.000000000 -0700
@@ -1730,10 +1730,6 @@ ext4_mb_regular_allocator(struct ext4_al
 		ac->ac_g_ex.fe_start = sbi->s_mb_last_start;
 		spin_unlock(&sbi->s_md_lock);
 	}
-
-	/* searching for the right group start from the goal value specified */
-	group = ac->ac_g_ex.fe_group;
-
 	/* Let's just scan groups to find more-less suitable blocks */
 	cr = ac->ac_2order ? 0 : 1;
 	/*
@@ -1743,6 +1739,12 @@ ext4_mb_regular_allocator(struct ext4_al
 repeat:
 	for (; cr < 4 && ac->ac_status == AC_STATUS_CONTINUE; cr++) {
 		ac->ac_criteria = cr;
+		/*
+		 * searching for the right group start
+		 * from the goal value specified
+		 */
+		group = ac->ac_g_ex.fe_group;
+
 		for (i = 0; i < EXT4_SB(sb)->s_groups_count; group++, i++) {
 			struct ext4_group_info *grp;
 			struct ext4_group_desc *desc;
